nixos/wireguard: setup interface with systemd-networkd

dguibert opened the pull request:
###### Motivation for this change

This patch make the wireguard interface more resilient and should fix #41874.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [X] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

dguibert committed:
systemd-networkd: add support for wireguard netdev.

Mic92 commented on nixos/modules/system/boot/networkd.nix:
```diff
@@ -315,6 +330,29 @@ let
+      type = types.addCheck (types.attrsOf unitOption) checkWireGuard;
+      description = ''
+        Each attribute in this set specifies an option in the
+        <literal>[WIREGUARD]</literal> section of the unit.  See
```
`[WireGuard]`

dguibert replied:
done.

Mic92 commented on nixos/modules/system/boot/networkd.nix:
```diff
@@ -315,6 +330,29 @@ let
+      type = with types; listOf (submodule wireguardPeerOptions);
+      description = ''
+        Each attribute in this set specifies an option in the
+        <literal>[WIREGUARDPEER]</literal> section of the unit.  See
```
`[WireGuardPeer]`

Mic92 commented on nixos/modules/system/boot/networkd.nix:
```diff
@@ -445,6 +483,23 @@ let
+        type = types.addCheck (types.attrsOf unitOption) checkWireGuardPeer;
+        description = ''
+          Each attribute in this set specifies an option in the
+          <literal>[Route]</literal> section of the unit.  See
```
`WireguardPeer`

Mic92 reviewed (approved):
I have not tested the actual module, but the transition makes sense.

Mic92 commented:
Now we would only need  `suppress_prefixlength 0` in `[RoutingPolicyRule]` to support routes for `0.0.0.0`

andir commented on nixos/modules/services/networking/wireguard.nix:
```diff
@@ -182,17 +182,41 @@ let
       {
         description = "WireGuard Tunnel - ${name}";
-        after = [ "network.target" ];
+        after = [ "sys-subsystem-net-devices-${name}.device" ];
```
shouldn't we wait for DNS/network configuration?

I am not sure what the behavior ended up being (@Mic92) within systemd.

dguibert replied:
Maybe we should.
(btw systemd handles the [resolution of endpoints](https://github.com/systemd/systemd/blob/master/src/network/netdev/wireguard.c#L273))

Mic92 replied:
I added exponential backoff for dns resolution, so it will retry failed endpoints just like openvpn:  https://github.com/systemd/systemd/blob/master/src/network/netdev/wireguard.c#L268

It make sense to not wait for dns to avoid leaking traffic.

andir commented:
OK :+1:

On Tue, 21 Aug 2018, 11:42 Jörg Thalheim, <notifications@github.com> wrote:

> *@Mic92* commented on this pull request.
> ------------------------------
>
> In nixos/modules/services/networking/wireguard.nix
> <https://github.com/NixOS/nixpkgs/pull/45392#discussion_r211543052>:
>
> >      nameValuePair "wireguard-${name}"
>        {
>          description = "WireGuard Tunnel - ${name}";
> -        after = [ "network.target" ];
> +        after = [ "sys-subsystem-net-devices-${name}.device" ];
>
> I added exponential backoff for dns resolution, so it will retry failed
> endpoints just like openvpn:
> https://github.com/systemd/systemd/blob/master/src/network/netdev/wireguard.c#L268
>
> It make sense to not wait for dns to avoid leaking traffic.
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/NixOS/nixpkgs/pull/45392#discussion_r211543052>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAm_dFj8W4i0BSZpFk5Hc6-qBHoqnrfbks5uS9X3gaJpZM4WEUxh>
> .
>

globin reviewed (changes requested):
This misses creating routes, which was done before.


Patch for the inline comment on privateKey below.
Also added back postShutdown using `ip set link down`. We might not want to do this in case we handle routes with networkd, too.

```patch
From 854ce74e6eb0e7362ec9a749bdda3a83b7343795 Mon Sep 17 00:00:00 2001
From: Robin Gloster <mail@glob.in>
Date: Wed, 22 Aug 2018 04:01:51 +0200
Subject: [PATCH] wireguard: fix privateKey handling, add back postShutdown

---
 nixos/modules/services/networking/wireguard.nix | 9 +++++++--
 1 file changed, 7 insertions(+), 2 deletions(-)

diff --git a/nixos/modules/services/networking/wireguard.nix b/nixos/modules/services/networking/wireguard.nix
index 43eecca1b6c..75301179706 100644
--- a/nixos/modules/services/networking/wireguard.nix
+++ b/nixos/modules/services/networking/wireguard.nix
@@ -225,14 +225,19 @@ let
         };
 
         script = ''
-          wg set ${name} private-key ${values.privateKeyFile}
+          ${optionalString (values.privateKeyFile != null) "wg set ${name} private-key ${values.privateKeyFile}"}
           ${concatMapStringsSep "\n" (peer:
-              optionalString (peer.presharedKeyFile != null) "wg set ${name} peer  preshared-key ${peer.presharedKeyFile}"
+              optionalString (peer.presharedKeyFile != null) "wg set ${name} peer preshared-key ${peer.presharedKeyFile}"
             ) values.peers}
 
           ip link set up dev ${name}
           ${values.postSetup}
         '';
+
+        preStop = ''
+          ip link set down dev ${name}
+          ${values.postShutdown}
+        '';
       };
 
 in
-- 
2.16.4
```

globin commented on nixos/modules/services/networking/wireguard.nix:
```diff
@@ -201,45 +225,14 @@ let
-          wg set ${name} private-key ${privKey} ${
-            optionalString (values.listenPort != null) " listen-port ${toString values.listenPort}"}
-
+          wg set ${name} private-key ${values.privateKeyFile}
```
This should be wrapped in `optionalString (values.privateKeyFile != null)` to handle the use of `privateKey`

dguibert replied:
sure.

dguibert commented on nixos/modules/system/boot/networkd.nix:
```diff
@@ -46,11 +46,26 @@ let
       "bond" "bridge" "dummy" "gre" "gretap" "ip6gre" "ip6tnl" "ip6gretap" "ipip"
       "ipvlan" "macvlan" "macvtap" "sit" "tap" "tun" "veth" "vlan" "vti" "vti6"
       "vxlan" "geneve" "vrf" "vcan" "vxcan" "wireguard" "netdevsim"
+      "wireguard"
```
I've missed that. Thanks for pointing this out.

dguibert commented:
@globin, I've applied your patch and restored the option allowedIPsAsRoutes.

globin commented:
Ah I have one more minor nitpick to the route handling that already had existed before this PR. When adding routes without specifying `proto`, this makes some dynamic routing daemons ignore them (e.g. babel, probably more). Citing from `ip(8)`:
> If the routing protocol ID is not  given,  ip  assumes  protocol boot (i.e. it assumes the route was added by someone who doesn't understand what they are doing).

Patch to add `proto static`:
```patch
From 9e699b21199d2d05025c069efe60afd03546742a Mon Sep 17 00:00:00 2001
From: Robin Gloster <mail@glob.in>
Date: Wed, 22 Aug 2018 19:10:32 +0200
Subject: [PATCH] wireguard: make wg routes proto static

This fixes distributing routes with e.g. babel that ignores `proto boot`
---
 nixos/modules/services/networking/wireguard.nix | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/nixos/modules/services/networking/wireguard.nix b/nixos/modules/services/networking/wireguard.nix
index f6d45c9d357..46fcb03d32f 100644
--- a/nixos/modules/services/networking/wireguard.nix
+++ b/nixos/modules/services/networking/wireguard.nix
@@ -234,7 +234,7 @@ let
 
           ${optionalString (values.allowedIPsAsRoutes != false) (concatStringsSep "\n" (concatMap (peer:
               (map (allowedIP:
-                "ip route replace ${allowedIP} dev ${name} table ${values.table}"
+                "ip route replace ${allowedIP} dev ${name} table ${values.table} proto static"
               ) peer.allowedIPs)
             ) values.peers))}
 
-- 
2.16.4
```

globin commented:
Also this will not work without having `networking.useNetworkd` enabled right? I don't think we can require people to use networkd with wireguard right now, there are still to many issues that need to be fixed..

Mic92 commented:
It is possible to use `systemd.network.enable` without using `networking.useNetworkd`. Networkd only cares about interfaces it was explicitly configured for.

globin commented:
Then this should mkDefault enable it :)

dguibert commented:
Commits have been fix up to use systemd.network.enable and keep the DEVICE env. variable in the unit...

I've noticed that the unit is not reloaded when the device is reconfigured with networkd. The privateKey is not set anymore on the wireguard link after "systemctl restart system-networkd".
Ideas how to fix it?

Mic92 commented:
Indeed, networkd will overwrite the key whenever it configures an interface https://github.com/systemd/systemd/blob/master/src/network/netdev/wireguard.c#L69

The only way this could be prevented would be to create an .netdev unit that has the private key included.
I think we should at least merge `nixos/modules/system/boot/networkd.nix` since they are independent from the module, I have no idea how to cleanly combine privateKeyFile and networkd module.

dguibert commented:
For sure it could be prevented when the privateKey (and presharedKey) is defined instead of privateKeyFile.
The last patch aims to solve the reloading issue.

yorickvP commented:
Does this make wireguard depend on systemd-networkd?!

Mic92 commented on nixos/modules/services/networking/wireguard.nix:
```diff
@@ -201,43 +230,30 @@ let
-          wg set ${name} private-key ${privKey} ${
-            optionalString (values.listenPort != null) " listen-port ${toString values.listenPort}"}
+	  # wait for ${name} to be configured
+	  while ! networkctl status $DEVICE | grep configured 1>/dev/null; do
```
Ugh, this is ugly. What if the users restart systemd-networkd manually?

globin replied:
Could this be fixed with PartOf=systemd-networkd.service?

Also the indentation is off in a few places.

Mic92 replied:
Maybe?

Mic92 replied:
Also then module should set the device up with `ip` after networkctl has configured the device. Does not networkd also handles that?

dguibert replied:
This could be fixed by waiting `systemd-network-wait-online@${name}.service`. How could we make this unit from `systemd-network-wait-online@.service`?

The while loop could be replaced with `${config.systemd.package}/lib/systemd/systemd-networkd-wait-online -i ${name}`. If systemd-networkd is restarted, the keys are restored as expected.

Mic92 commented on nixos/modules/services/networking/wireguard.nix:
```diff
@@ -182,17 +182,46 @@ let
       {
         description = "WireGuard Tunnel - ${name}";
-        after = [ "network.target" ];
+        after = [ "sys-subsystem-net-devices-${name}.device" "systemd-networkd-online" ];
```
The target is spelled wrong and I don't see why `systemd-networkd-online.target` is required.

globin replied:
Also use `utils.escapeSystemdPath` on name in the unit definition as systemd will escape e.g. hyphens with \x2d

Mic92 commented on nixos/modules/services/networking/wireguard.nix:
```diff
@@ -201,43 +230,30 @@ let
+	  while ! networkctl status $DEVICE | grep configured 1>/dev/null; do
+	    sleep 1
+	  done
+          sleep 1
```
Why this sleep?

yorickvP commented on nixos/modules/services/networking/wireguard.nix:
```diff
@@ -201,43 +230,30 @@ let
         };
 
         script = ''
-          modprobe wireguard
```
Do you load the module anywhere?

xeji commented:
I merged #45569 because I am not sure this is ready yet, and we should get at least one of these improvements into 18.09. The merge does *not* imply a vote against this change.
It just needs a trivial rebase. Sorry for the inconvenience.

d6e commented:
@dguibert Is it just the merge conflict that's blocking this? Is there any way I can help?

dguibert commented:
Just rebased. Works for me since I've pushed the first patch series.
Is there still anything blocking this to be merged?

arianvp commented:
Heya. I recently got a patch merged in systemd that adds a `KeyFile` option to the netdev spec https://github.com/systemd/systemd/pull/11861

asymmetric commented:
[This](https://github.com/systemd/systemd/pull/11861/commits/76df77794a902c7a7bce6cf689114fde9e139b22) is the specific commit in the PR that makes the change, btw.

dguibert commented:
> Heya. I recently got a patch merged in systemd that adds a `PrivateKeyFile` option to the netdev spec [systemd/systemd#11861](https://github.com/systemd/systemd/pull/11861) . However we do not ship this yet but would be nice to add that to the future instead of the scripts.

Great! I've just updated to whole branch with the last commit supporting this new option.

PS: Maybe a presharedKeyFile option could be also added to systemd ;-)

picnoir committed:
systemd-networkd: add tests

(cherry picked from commit ec073e41a0dc8273cd81cf61fa37004310120af2)

zarelit commented:
How is this PR related to https://github.com/NixOS/nixpkgs/pull/64040 ? Should we keep both / merge them / close one?

dguibert commented:
PR #64040 introduces wireguard related options as well as this PR, but there are some differences.
- This PR converts the options config.networking.wireguard to networkd options, this part could be removed as it conflicts with some committed changes.
- It supports multiple peers per interface (as far as I can tell not the new PR).

NinjaTrappeur commented:
> It supports multiple peers per interface (as far as I can tell not the new PR).

It does support it.

If you prefer merging this one, please do add the nixos test I wrote in #64040 .

dguibert commented:
> > It supports multiple peers per interface (as far as I can tell not the new PR).
> 
> It does support it.

I don't figure it out. I've tried to add this to the test

`
wireguardPeerConfig = {
    Endpoint = "192.168.1.3:51820";
    PublicKey = pubk;
    PresharedKeyFile = pkgs.writeText "psk.key" "yTL3sCOL33Wzi6yCnf9uZQl/Z8laSE+zwpqOHC4HhFU=";
    AllowedIPs = [ "10.0.0.3/32" ];
    PersistentKeepalive = 15;
};
`

but `error: attribute 'wireguardPeerConfig' at /nixpkgs/nixos/tests/systemd-networkd-wireguard.nix:26:13 already defined at /nixpkgs/nixos/tests/systemd-networkd-wireguard.nix:19:13`

> If you prefer merging this one, please do add the nixos test I wrote in #64040 .

done.

NinjaTrappeur commented:
Thanks @dguibert. Closed the other one.

My bad for the multi-peer faulty behavior.

Looking forward to see this PR merged.

flokli reviewed (approved):
LGTM, thanks!
